src/Microsoft.Net.Sdk.AnalyzerRedirecting/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,142 @@
+
+// Example:
+// FullPath: "C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\8.0.8\analyzers\dotnet\System.Windows.Forms.Analyzers.dll"
+// ProductVersion: "8.0.8"
+// PathSuffix: "analyzers\dotnet"
+using AnalyzerInfo = (string FullPath, string ProductVersion, string PathSuffix);
+
+namespace Microsoft.Net.Sdk.AnalyzerRedirecting;
+
+[Export(typeof(IAnalyzerAssemblyRedirector))]
+public sealed class SdkAnalyzerAssemblyRedirector : IAnalyzerAssemblyRedirector
+{
+    private readonly string? _insertedAnalyzersDirectory;
+    private readonly Lazy<ImmutableDictionary<string, List<AnalyzerInfo>>> _analyzerMap;
+
+    [ImportingConstructor]
+    public SdkAnalyzerAssemblyRedirector()
+        : this(Path.GetFullPath(Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "..", "..", "SDK", "RuntimeAnalyzers"))) { }
+
+    // Internal for testing.
+    internal SdkAnalyzerAssemblyRedirector(string? insertedAnalyzersDirectory)
+    {
+        _insertedAnalyzersDirectory = insertedAnalyzersDirectory;
+        _analyzerMap = new(CreateAnalyzerMap);
+    }
+
+    /// <summary>
+    /// Map from analyzer assembly name (file name without extension) to a list of matching analyzers.
+    /// </summary>
+    private ImmutableDictionary<string, List<AnalyzerInfo>> AnalyzerMap => _analyzerMap.Value;
+
+    private ImmutableDictionary<string, List<AnalyzerInfo>> CreateAnalyzerMap()
+    {
+        var builder = ImmutableDictionary.CreateBuilder<string, List<AnalyzerInfo>>(StringComparer.OrdinalIgnoreCase);
+
+        // Expects layout like:
+        // VsInstallDir\SDK\RuntimeAnalyzers\WindowsDesktopAnalyzers\8.0.8\analyzers\dotnet\System.Windows.Forms.Analyzers.dll
+        //                                   ~~~~~~~~~~~~~~~~~~~~~~~                                                           = topLevelDirectory
+        //                                                           ~~~~~                                                     = versionDirectory
+        //                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ = analyzerPath
+
+        foreach (string topLevelDirectory in Directory.EnumerateDirectories(_insertedAnalyzersDirectory))
+        {
+            foreach (string versionDirectory in Directory.EnumerateDirectories(topLevelDirectory))
+            {
+                foreach (string analyzerPath in Directory.EnumerateFiles(versionDirectory, "*.dll", SearchOption.AllDirectories))
+                {
+                    if (!analyzerPath.StartsWith(versionDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    string version = Path.GetFileName(versionDirectory);
+                    string analyzerName = Path.GetFileNameWithoutExtension(analyzerPath);
+                    string pathSuffix = analyzerPath.Substring(versionDirectory.Length + (EndsWithSlash(versionDirectory) ? 0 : 1));
```

Comment:
Shouldn't the Directory API either always give you a `slash` or not?  Why do you need to handle both cases?

Reply:
This does not seem to be guaranteed by the docs, but I guess we can rely on that and simplify the code, thanks.